File: RxHttp/src/main/java/com/github/kongpf8848/rxhttp/typebuilder/TypeToken.java
package com.github.kongpf8848.rxhttp.typebuilder;

import com.github.kongpf8848.rxhttp.typebuilder.exception.TypeException;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;


public abstract class TypeToken<T> {
    private final Type type;

    public TypeToken() {
        Type superclass = getClass().getGenericSuperclass();
        if (superclass instanceof Class) {
            throw new TypeException("No generics found!");
        }
        ParameterizedType type = (ParameterizedType) superclass;
        this.type = type.getActualTypeArguments()[0];
    }

    public Type getType() {
        return type;
    }
}

File: RxHttp/src/main/java/com/github/kongpf8848/rxhttp/typebuilder/TypeToken.kt
package com.github.kongpf8848.rxhttp.typebuilder

import com.github.kongpf8848.rxhttp.typebuilder.exception.TypeException
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

abstract class TypeToken<T> {

    val type: Type

    init {
        val superclass = javaClass.genericSuperclass
        if (superclass is Class<*>) {
            throw TypeException("No generics found!")
        }
        val type = superclass as ParameterizedType
        this.type = type.actualTypeArguments[0]
    }
}